internal/pkg/providers/logging_provider.go

```diff
@@ -0,0 +1,29 @@
+package providers
+
+import (
+	"gses4_project/internal/services"
+	"log"
+)
+
+type (
+	LoggingProvider struct {
+		name         string
+		rateProvider services.IRateAPIProvider
+	}
+)
+
+func NewLoggingProvider(name string, rateProvider services.IRateAPIProvider) *LoggingProvider {
+	return &LoggingProvider{
+		name:         name,
+		rateProvider: rateProvider,
+	}
+}
+func (l *LoggingProvider) FetchRate() (*float64, error) {
+	rate, err := l.rateProvider.FetchRate()
+	if err != nil {
+		log.Printf("%v provider returned error: %v", l.name, err)
+		return nil, err
+	}
```

Comment:
why do we need to fetch rate in the logging provider? 
maybe it depends on approach, but would not it be better to pass logger to the rate provider and not vice versa?

Reply:
Here I've decided to decorate each provider with the logger so I don't need to add logs in future implementations, I will just decorate it with logger and get my logs